include/boost/url/grammar/parse.hpp

```diff
@@ -77,216 +63,166 @@ parse(
+    @code
+    template< class Rule >
+    bool parse( char const*&, char const*, error_code&, Rule&& r );
+    @endcode
 
-   The function returns `true` and resets @ref ec when the input is
-   parsed successfully. If the input cannot be parsed into the BNF elements,
-   the function returns `false` and sets the error number in @ref ec to the
-   appropriate error number defined in @ref boost::urls::error.
+    If this call succeeds, the algorithm proceeds by
+    performing the same operation on the remaining
+    arguments including the parameter pack.
 
-   @par Example
-   @code
-   #include <iostream>
-   #include <string>
-
-   #include <boost/url/grammar/char_set.hpp>
-   #include <boost/url/grammar/parse.hpp>
-   #include <boost/url/error.hpp>
-
-   namespace boost {
-   namespace urls {
-   namespace detail {
-   // BNF element to parse a non-empty sequence of alpha chars ([a-zA-Z]+)
-   struct alpha_string_rule {
-     // reference to a std::string where we should store the results
-     std::string &v;
-
-     // parsing an alpha_string_rule element
-     friend bool parse(char const *&it, char const *end, error_code &ec,
-                       alpha_string_rule &&t) {
-       // is_alpha functor
-       grammar::alpha_chars is_alpha;
-
-       // check if empty
-       if (it == end) {
-         ec = error::incomplete;
-         return false;
-       }
-
-       // check if not alpha at all
-       if (!is_alpha(*it)) {
-         ec = error::bad_alpha;
-         return false;
-       }
-       t.v.push_back(*it);
-       ++it;
-
-       // consume more alphas
-       while (it != end && is_alpha(*it)) {
-         t.v.push_back(*it);
-         ++it;
-       }
-       ec = error::success;
-       return true;
-     }
-   };
-   } // namespace detail
-   } // namespace urls
-   } // namespace boost
-
-   int main() {
-     using namespace boost::urls;
-
-     // Input
-     std::string str = "Hello.cpp";
-     auto it = str.data();
-     auto const end = it + str.size();
-
-     // Output
-     std::string alpha_l;
-     std::string alpha_r;
-     error_code ec;
-
-     // Parse two alpha strings separated by a '.'
-     if (grammar::parse(it, end, ec, detail::alpha_string_rule{alpha_l}, '.',
-                    detail::alpha_string_rule{alpha_r})) {
-       std::cout << alpha_l << std::endl;
-       std::cout << alpha_r << std::endl;
-     }
-     return 0;
-   }
-   @endcode
+    If an error occurs for any of the rules in the
+    parameter pack, the entire parse operation fails
+    but the chars from the substring are still consumed.
 
-   @par Exception Safety
-     Calls to `parse` overloads may throw, although they should
-     prefer setting @ref ec instead.
+    As other overloads of this function, `using grammar::parse`
+    is required to call this function appropriately with for
+    argument-dependent lookup for other grammar rules, and
+    calls to `parse` should be unqualified.
```

Comment:
"As other overloads of..." doesn't sound right. How about:

> The statement `using grammar::parse` should be in scope before any calls to `parse` are made without namespace qualification, so that argument-dependent lookup can take effect, as seen in this example:

Reply:
Better